designs/national-cloud-support.md

```diff
@@ -0,0 +1,104 @@
+
+#### Authentication
+
+Authentication is also reading from [/azure/config/azure.json](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/azclient/auth_conf.go#L27)
+
+
+#### Decision
+
+**Pros**
+
+* Existing helpers to deserialize well-known formats like `AZURE_ENVIRONMENT_FILEPATH` (originally from track1), `/azure/config/azure.json`.
+* Client [ratelimit support](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/azclient/policy/ratelimit/ratelimit.go#L46) - configured from `/azure/config/azure.json` by default, but disabled by default.
+
+**Cons**
+
+* Non-idiomatic [factory interface](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/azclient/factory.go) for use with GoMock.
+* Cloud environment setup is honestly a big mess. A lot of ways to configure things, various inputs from multiple files that can be ignored or overwritten (or partially ignored) in certain cases.
+  Doesn't actually map all of the values we would need now so we would need to contribute upstream to get graph endpoint mapped to track2 configuration.
+* Doesn't expose the full featureset of the underlying SDK - for example [no Pagination on VM list](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/azclient/virtualmachineclient/interface.go#L36).
+  We may not need these features but if we ever do, additional overhead to add them.
+* We already do authentication differently (via env variables), passing `--cloud-config` and taking `/azure/config/azure.json` would be confusing because now we need to answer which wins, the auth details in that file or the
+  auth details we accept already via the environment?
+* It would likely be a significant amount of work to change to the CloudProvider clients. The benefits don't seem worth the engineering investment.
+
+**Conclusion:** No, we should continue using the raw clients. We _may_ want to refer to some helpers from CloudProvider for file structure and/or utilities like rate-limiting, but those can be plugged directly into the raw Go SDK without
+needing to use the full CloudProvider client set.
+
+### How should we pass the cloud details to Karpenter?
+
+The main disadvantage of using `/azure/config/azure.json` is that it is a "bag of everything". Introducing it into
+Karpenter configuration might cause confusion about exactly which fields are we using from it. Do we honor the throttling details? Can users configure client throttling separately for Karpenter vs CloudProvider?
+
+On the other hand, `/AzureStackCloud/environment.json`, often passed to the `AZURE_ENVIRONMENT_FILEPATH` env variable, is singularly focused on describing the cloud environment, which is what we want and wouldn't add confusion.
+It has the advantage of being on the nodes already in clouds that require it, saving users from needing to configure the endpoints themselves in the Karpenter chart in selfhosted scenarios.
```

Comment:
Need to confirm exactly where this is written from to make sure this statement is true in all contexts.

Reply:
Confirmed this is written by [AgentBaker](https://github.com/Azure/AgentBaker/blob/d786bf15845160db41fcb92f45928f290439582c/aks-node-controller/parser/parser.go#L157), but it's written to `/etc/kubernetes/AzureStackCloud.json` not `/AzureStackCloud/environment.json` like I had originally said.

Corrected the document and included the above link.